Sync Local Pickup title between Checkout block and shipping settings UI and vice/versa

opr opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

This PR introduces some changes that will cause the local pickup title to save to the settings UI when edited in the block, and save to the block when edited in the settings UI.

To facilitate the sync from settings -> editor/front-end, the current local pickup title is output in the asset data registry (accessible via `wc.wcSettings`) this value is used by the block to render the title.

To handle syncing from editor -> settings UI, the `post_saved` action is hooked and we check if the post being saved is the checkout template, or the checkout page. If so, then the title text from the `Local Pickup` selector is extracted and saved to the settings.

This PR also adds E2E tests which checks the sync between editor/frontend and settings UI and vice/versa.

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Closes #45160.

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Use a block theme such as Twenty Twenty-Four. 
2. Go to WooCommerce -> Settings -> Shipping -> Local Pickup.
3. Add a location and enable local pickup.
4. Change the title of local pickup to something different.
5. Open the Checkout page template in the Site editor.
6. Ensure you can see the updated local pickup title from step 4.
<img width="538" alt="image" src="https://github.com/woocommerce/woocommerce/assets/5656702/76857093-a768-4699-a2f0-f5a9355d5dee">

7. Click into the local pickup selector and edit the title to something different.
8. Save the template and then go back to WooCommerce -> Settings -> Shipping -> Local Pickup.
9. Ensure the title shown here is the same as what you entered in the site editor.

https://www.loom.com/share/9324af6dd240411c839e944ac09ae367

1. Use a classic theme, such as storefront.
2. Repeat the steps above, but instead of going to the site editor, go to Pages -> then to your checkout page.

#### Edge case testing

1. Add a new page - insert the checkout block and edit the local pickup title.
2. Ensure it does **not** update the local pickup title in the settings UI. (this is because it only applies to the designated Checkout page).

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

#### Comment <!-- If the changes in this pull request don't warrant a changelog entry, you can alternatively supply a comment here. Note that comments are only accepted with a significance of "Patch" -->

</details>

opr committed:
Add localPickupText to asset_data_registry on checkout block

opr committed:
Add find_local_pickup_text_in_checkout_block function

This recurses through the inner blocks and tries to find the local pickup text.

opr committed:
Add update_local_pickup_title that runs on save

This will update the core setting when the local pickup text is changed.

opr committed:
Get localPickupText from wcSettings in frontend block

opr committed:
Update localPickupText attribute from wcSettings in editor

opr committed:
Add defensive checks and handle site editor and page editor

opr committed:
Add setTitle function to localPickupUtils

opr committed:
Wait for snackbar confirmation when saving local pickup settings

opr committed:
Add e2e test for local pickup title setting

opr committed:
Add changelog

opr committed:
Remove unused import

opr committed:
Fix lint errors

opr committed:
Fix further lint errors

nielslange commented on plugins/woocommerce/src/Blocks/BlockTypes/Checkout.php:
```diff
@@ -242,6 +244,71 @@ protected function is_checkout_endpoint() {
+
+		if ( ( ! empty( $post->post_type ) && ! empty( $post->post_name ) && 'page-checkout' !== $post->post_name && 'wp_template' === $post->post_type ) || false === has_block( 'woocommerce/checkout', $post ) ) {
+			return;
+		}
+		$pickup_location_settings = get_option( 'woocommerce_pickup_location_settings', array() );
```
_When in doubt, space it out._ 😁

```suggestion
		}
		
		$pickup_location_settings = get_option( 'woocommerce_pickup_location_settings', array() );
```

nielslange commented on plugins/woocommerce/src/Blocks/BlockTypes/Checkout.php:
```diff
@@ -242,6 +244,71 @@ protected function is_checkout_endpoint() {
+
+		$post_blocks = parse_blocks( $post->post_content );
+		$title       = $this->find_local_pickup_text_in_checkout_block( $post_blocks );
+		if ( $title ) {
```
Adding a line break here as well, might also increase the readability.

```suggestion
		$title       = $this->find_local_pickup_text_in_checkout_block( $post_blocks );
		
		if ( $title ) {
```

nielslange commented on plugins/woocommerce/src/Blocks/BlockTypes/Checkout.php:
```diff
@@ -242,6 +244,71 @@ protected function is_checkout_endpoint() {
+	 * Recurse through the blocks to find the shipping methods block, then get the value of the localPickupText attribute from it.
+	 *
+	 * @param array $blocks The block(s) to search for the local pickup text.
+	 * @return void|string  The local pickup text if found, otherwise void.
```
Should the `@return` annotation here be `null|string` instead? Typically, `@return void` implies that the method doesn't return a value. However, in the described method, it either returns the `localPickupText` attribute's value or it returns `null` when no value is specified.

nielslange commented on plugins/woocommerce-blocks/assets/js/blocks/checkout/inner-blocks/checkout-shipping-method-block/block.tsx:
```diff
@@ -172,6 +172,10 @@ const Block = ( {
 		'shippingCostRequiresAddress',
 		false
 	);
+	const localPickupTextFromSettings = getSetting< string >(
+		'localPickupText',
+		localPickupText || defaultLocalPickupText
+	);
```
Nitpic: I wonder if we should use `localPickupTitle` and `defaultLocalPickupTitle` instead of `localPickupText` and `defaultLocalPickupText`. 

1. In the editor, we call that field title.
<img width="458" alt="Screenshot 2024-03-27 at 13 33 10" src="https://github.com/woocommerce/woocommerce/assets/3323310/98e137ae-6729-4983-99dd-1cab64b44f65">

2. In the frontend, the top text looks like the title, while the bottom text looks like regular text.
<img width="350" alt="Screenshot 2024-03-27 at 13 33 24" src="https://github.com/woocommerce/woocommerce/assets/3323310/8880af1b-8ea2-4920-acea-b01bb136c823">

If we'd adjust that, we'd have to change it in various areas. I believe such a change would make the code a bit easier to read, but I'll leave that up to you.

opr replied:
Thanks for the suggestion, it makes sense but I feel this is out of scope for this PR and changing it would increase testing. We could address that in a refactor in future.

nielslange commented on plugins/woocommerce-blocks/tests/e2e/tests/local-pickup/local-pickup.merchant.block_theme.side_effects.spec.ts:
```diff
@@ -0,0 +1,96 @@
+				details: 'American entity',
+			},
+		} );
+		await localPickupUtils.disableLocalPickupCosts();
```
As we're not checking against local pickup costs, do we need this part?

nielslange commented on plugins/woocommerce-blocks/tests/e2e/tests/local-pickup/local-pickup.merchant.block_theme.side_effects.spec.ts:
```diff
@@ -0,0 +1,96 @@
+
+	test.afterEach( async ( { localPickupUtils } ) => {
+		await localPickupUtils.deleteLocations();
+		await localPickupUtils.disableLocalPickupCosts();
```
Same as my previous comment. I wonder if we need this part.

nielslange commented on plugins/woocommerce-blocks/tests/e2e/tests/local-pickup/local-pickup.merchant.block_theme.side_effects.spec.ts:
```diff
@@ -0,0 +1,96 @@
+		await localPickupUtils.deleteLocations();
+		await localPickupUtils.disableLocalPickupCosts();
+		await localPickupUtils.setTitle( 'Local Pickup' );
+		await localPickupUtils.enableLocalPickup();
```
In `test.beforeEach`, we're enabling local pickups. In the tests, I do not see that we're disabling them. Do we really need to enable them again in `test.afterEach`? I'd assume that they're still enabled.

opr replied:
We basically want to reset the local pickup to how it was before we saw it, but since we need a page object to do so, we can't do this in `beforeAll` or `afterAll`, therefore I defaulted to `beforeEach` and `afterEach` where page contexts exist

nielslange commented on plugins/woocommerce-blocks/tests/e2e/utils/local-pickup/local-pickup-utils.page.ts:
```diff
@@ -32,6 +33,17 @@ export class LocalPickupUtils {
 
 	async saveLocalPickupSettings() {
 		await this.page.getByRole( 'button', { name: 'Save changes' } ).click();
+		await this.page.waitForFunction( () => {
+			return window.wp.data
+				.select( 'core/notices' )
+				.getNotices()
+				.some(
+					( notice: Notice ) =>
+						notice.status === 'success' &&
+						notice.content ===
+							'Local Pickup settings have been saved.'
+				);
+		} );
```
In our e2e tests, we don't seem to check against this notice content. Do we really need this part? If so, shouldn't we handle the notice content?

opr replied:
Sorry Niels but I don't really follow. This check is in place to see when the save action has completed. We could check against request/response completion too but this seemed easier.

nielslange replied:
What I mean is, that we're not checking if the notice `Local Pickup settings have been saved.` appears, did we? At least, I could not see it. That's why I was asking if this part is necessary.

opr replied:
Thanks for the suggestion, I've updated the test to look for the text now, removed the data store check.

nielslange replied:
Thanks for adjusting this, Thomas.

nielslange commented on plugins/woocommerce-blocks/tests/e2e/utils/local-pickup/local-pickup-utils.page.ts:
```diff
@@ -60,6 +72,12 @@ export class LocalPickupUtils {
 		await this.saveLocalPickupSettings();
 	}
 
+	async setTitle( title: string ) {
+		await this.openLocalPickupSettings();
+		await this.page.getByLabel( 'Title' ).fill( title );
+		await this.saveLocalPickupSettings();
+	}
```
What are your thoughts on:
- renaming this function from `setTitle` to `setLocalPickupTitle` to be consistent with the other function names
- moving this function below `disableLocalPickupCosts()` as it's currently between `enableLocalPickupCosts()` and `disableLocalPickupCosts()` to keep these functions together

opr replied:
Nice idea, thanks for the suggestion

nielslange commented on plugins/woocommerce/src/Blocks/BlockTypes/Checkout.php:
```diff
@@ -242,6 +244,71 @@ protected function is_checkout_endpoint() {
+	 * @return void
+	 */
+	public function update_local_pickup_title( $post_id, $post ) {
+		if ( empty( $post->post_status ) || 'inherit' === $post->post_status ) {
+			// This is not a proper save action, maybe an autosave, so don't continue.
+			return;
+		}
```
What are your thoughts on moving the comment above the conditional check?

```suggestion
		// This is not a proper save action, maybe an autosave, so don't continue.
		if ( empty( $post->post_status ) || 'inherit' === $post->post_status ) {
			return;
		}
```

nielslange reviewed (approved):
Thanks for working on this, @opr. The PR works as expected and the code looks great. Well done. 👏

I just left a few minor comments and suggestions, but pre-approving this PR as it's already in a great shape.

opr committed:
Update function return type to be explicitly null

opr committed:
Add spacing

opr committed:
Remove unnecessary repeated e2e teardown tasks

opr committed:
Rename setTitle to setLocalPickupTitle in e2e utils

opr committed:
Enter edit mode in canvas instead of clicking

opr committed:
Move comment above condition

opr commented:
@nielslange thanks for your review, I addressed the comments, but I don't think there's anything actionable from this one: https://github.com/woocommerce/woocommerce/pull/45720#discussion_r1540559065

On this basis I've enabled auto merge for this PR.

opr committed:
Wait for notice text to be visible instead of checking data store

opr committed:
Fix linting errors
